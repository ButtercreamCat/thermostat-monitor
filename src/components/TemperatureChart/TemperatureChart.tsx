import React, { Component } from "react";
import FusionCharts from "fusioncharts";
import Line from "fusioncharts/fusioncharts.zoomline";
import ReactFC from "react-fusioncharts";
import FusionTheme from "fusioncharts/themes/fusioncharts.theme.fusion";
import axios from "axios";

ReactFC.fcRoot(FusionCharts, Line, FusionTheme);

interface TemperatureChartState {
  chart: {
    caption: string;
    subcaption: string;
    yaxisname: string;
    xaxisname: string;
    numberSuffix: string;
    forceaxislimits: string;
    pixelsperpoint: string;
    pixelsperlabel: string;
    compactdatamode: string;
    dataseparator: string;
    bgColor: string;
    theme: string;
  };
  categories: { category: string }[] | undefined;
  dataset: { seriesname: string; data: string }[] | undefined;
}

class TemperatureChart extends Component<{}, TemperatureChartState> {
  state: TemperatureChartState = {
    chart: {
      caption: "Temperature",
      subcaption: "Click & drag on the plot area to zoom & then scroll",
      yaxisname: "Temperature",
      xaxisname: "Date",
      numberSuffix: "Â°C",
      forceaxislimits: "1",
      pixelsperpoint: "0",
      pixelsperlabel: "30",
      compactdatamode: "1",
      dataseparator: "|",
      bgColor: "#fafafa",
      theme: "fusion"
    },
    categories: undefined,
    dataset: undefined
  };

  componentDidMount() {
    axios
      .get(
        "https://raw.githubusercontent.com/ParityInc/backend-assignment/master/thermostat.json"
      )
      .then(response => {
        const pointData: {
          id: string;
          name: string;
          unit: string;
          graph_data: { x: string; actual: number }[];
        }[] = response.data.point_data;
        // Convert to fusionchart data format

        // Aggregate data for x-axis
        const categoriesReducer = (
          accumulator: string,
          currentValue: { x: string; actual: number }
        ) => {
          const trimmedString = currentValue.x
            .replace(":00-04:00", "")
            .replace("T", " - ");
          accumulator += trimmedString + "|";
          return accumulator;
        };
        const categoryResult = pointData[0].graph_data.reduce(
          categoriesReducer,
          ""
        );
        const categoriesResult = [
          {
            category: categoryResult
          }
        ];

        // Aggregate data for y-axis
        const datasetReducer = (
          accumulator: { seriesname: string; data: string }[],
          currentValue: {
            id: string;
            name: string;
            unit: string;
            graph_data: { x: string; actual: number }[];
          }
        ) => {
          const seriesObj: { seriesname: string; data: string } = {
            seriesname: "",
            data: ""
          };
          seriesObj.seriesname = currentValue.name;
          const temperatureReducer = (
            accumulator: string,
            currentValue: { x: string; actual: number }
          ) => {
            accumulator += currentValue.actual + "|";
            return accumulator;
          };
          seriesObj.data = currentValue.graph_data.reduce(
            temperatureReducer,
            ""
          );
          const result = accumulator.concat([seriesObj]);
          return result;
        };
        const datasetResult = pointData.reduce(datasetReducer, []);

        // Set State to trigger the re-render
        this.setState({
          categories: categoriesResult,
          dataset: datasetResult
        });
      })
      .catch(error => {
        console.log(error);
      });
  }

  render() {
    return (
      <div>
        {this.state.categories ? (
          <ReactFC
            type="zoomline"
            width="100%"
            height="100%"
            dataFormat="JSON"
            dataSource={this.state}
          />
        ) : (
          <div>Loading...</div>
        )}
      </div>
    );
  }
}

export default TemperatureChart;
