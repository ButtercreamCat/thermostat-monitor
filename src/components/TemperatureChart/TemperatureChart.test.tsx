import React from "react";
import { shallow, ShallowWrapper } from "enzyme";
import TemperatureChart from "./TemperatureChart";

jest.mock("axios", () => {
  const response = {
    data: {
      date: "30 May - 31 May",
      point_data: [
        {
          id: "ambient_temperature",
          name: "Ambient Temperature",
          unit: "°C",
          graph_data: [
            {
              actual: 23.42,
              x: "2019-05-30T00:00:00-04:00"
            }
          ]
        },
        {
          id: "outdoor_temperature",
          name: "Outdoor Temperature",
          unit: "°C",
          graph_data: [
            {
              actual: 13.03,
              x: "2019-05-30T00:00:00-04:00"
            }
          ]
        },
        {
          id: "target_temperature",
          name: "Target Temperature",
          unit: "°C",
          graph_data: [
            {
              actual: 25.5,
              x: "2019-05-30T00:00:00-04:00"
            }
          ]
        }
      ],
      report_unit: "fifteen_minutes"
    }
  };

  return {
    get: jest.fn(() => Promise.resolve(response))
  };
});

let wrapper: ShallowWrapper<any, Readonly<{}>, React.Component<{}, {}, any>>;
const axios = require("axios");

beforeEach(() => {
  wrapper = shallow(<TemperatureChart></TemperatureChart>);
});

test("should render TemperatureChart correctly", () => {
  expect(wrapper).toMatchSnapshot();
});

test("fetch data on componentDidMount", async () => {
  const instance = wrapper.instance() as TemperatureChart;
  instance.componentDidMount();
  const getSpy = jest.spyOn(axios, "get");
  expect(getSpy).toHaveBeenCalled();
  expect(getSpy).toHaveBeenCalledWith(
    "https://raw.githubusercontent.com/ParityInc/backend-assignment/master/thermostat.json"
  );
  await Promise.resolve();
  expect(wrapper.state()).toHaveProperty("categories", [
    { category: "2019-05-30 - 00:00|" }
  ]);
  expect(wrapper.state()).toHaveProperty("dataset", [
    { seriesname: "Ambient Temperature", data: "23.42|" },
    { seriesname: "Outdoor Temperature", data: "13.03|" },
    { seriesname: "Target Temperature", data: "25.5|" }
  ]);
});
