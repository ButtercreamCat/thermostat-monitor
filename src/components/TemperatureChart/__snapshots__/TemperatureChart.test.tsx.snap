// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render TemperatureChart correctly 1`] = `
<div>
  <ReactFC
    dataFormat="JSON"
    dataSource={
      Object {
        "categories": Array [
          Object {
            "category": "2019-05-30 - 00:00|",
          },
        ],
        "chart": Object {
          "bgColor": "#fafafa",
          "caption": "Temperature",
          "compactdatamode": "1",
          "dataseparator": "|",
          "forceaxislimits": "1",
          "numberSuffix": "Â°C",
          "pixelsperlabel": "30",
          "pixelsperpoint": "0",
          "subcaption": "Click & drag on the plot area to zoom & then scroll",
          "theme": "fusion",
          "xaxisname": "Date",
          "yaxisname": "Temperature",
        },
        "dataset": Array [
          Object {
            "data": "23.42|",
            "seriesname": "Ambient Temperature",
          },
          Object {
            "data": "13.03|",
            "seriesname": "Outdoor Temperature",
          },
          Object {
            "data": "25.5|",
            "seriesname": "Target Temperature",
          },
        ],
      }
    }
    height="100%"
    type="zoomline"
    width="100%"
  />
</div>
`;
